src/mode.js

```diff
@@ -121,7 +121,8 @@ function getToken(stream, state) {
       }
 
       // Otherwise, match a Terminal.
-      if (expected.match && expected.match(token)) {
+      const invariant = expected.constraints && expected.constraints(token);
+      if (!invariant && expected.match && expected.match(token)) {
```

Comment:
I'd suggest a name other than `invariant` - usually that's the name of a function. How about `areConstraintsMet`